### DISCUSSION TITLE
Sound and PTFX Online Sync

### DISCUSSION THREAD

**Discussion by xesdoog** [YimMenu/YimMenu#3351] [Q&A] [answered]
I have a question regarding spawned particle effects and sounds.
Using Lua, if you spawn PTFX effects on, let's say your car, like flames coming out of your exhaust, I've noticed that some assets were synchronized online so other players were able to see them and others weren't. I couldn't see any obvious reason for it other than the asset wasn't loaded in other players' machines so only we (the ones who spawned it) could see it?
Another question I have is about sounds started from Lua scripts. For example if I call `PLAY_SOUND_FROM_ENTITY()` or `PLAY_SOUND_FROM_COORD()` and for example make a prop play some sound, only I was able to hear it.

My question is are these two cases related? And is there anything I can do to sync these sounds and ptfx assets?

**Comment by ghost** [comment]
two clowns meet again, how cute

**Reply by ghost** [reply]
> > two clowns meet again, how cute
> 
> First of all, bitch who the fuck are you? And why are you commenting here? Second, you're definitely mistaking me for someone else because this is the first time I've ever had an argument with gir489. He's a damn good programmer, he's just an asshole and we all know that. What the fuck does that have to do with you?
> 
> Let me make some things clear to anyone else reading these comments: YimMenu's community is not hostile. As a new Lua dev, I've had conversations with several other developers and they are very helpful and friendly including gir489, he's just a fucking asshole but he helps when he feels like it.
> 
> This cringe arab cunt is just a cunt so ignore him.

Wow, what a long text. Apparently you are very retarded. If you take everything so personally, you will have a miserable life.

**Comment by Yimura** [comment]
Can we not do this, I swear you’re all acting like kindergarteners…

**Reply by xesdoog** [reply]
I was in a very bad mood yesterday. I apologize.

**Comment by FIying-Scotsman** [comment, marked as answer]
Back on topic - I noticed similar years ago (we're talking late 2013 - 2014) and didn't look any further into it other than most SP specific assets (PTFX and sounds) didn't sync, but ones which were used online did. There were a few outliers however; for example the clown smoke synced before it was used in certain races to transform your vehicle. We're talking 2014 ish which is well before transform races were even a thing. You can look through the list of PTFX dicts using tools like [this](https://vespura.com/fivem/particle-list/). Worth noting - even in SP there's dicts with particles which don't work either but I haven't looked at whether those are actually used or just left overs.

Best guess at the time was there's a list of effect dictionaries which can be loaded by online scripts, and those PTFX/sounds worked because they were in those same dicts (even though different effects within those dicts are meant to be used). If that's the case, it's likely at a sync level they're being filtered out for whatever reason so even if you send the correct sync - the other clients don't know how to action the request (i.e load this effect dict), so it's just dropped and it moves to the next packet.

If there is a system like that in place - you won't be able to force it on vanilla players because they're missing the logic to say "if this appears, do this with it". Mod menu users however might be able to have a background hook/patch to run custom code to handle those events so they sync for those players similar to 'menu chat' where you can hook the chat system and have menu <-- --> menu conversations as it's otherwise hidden from other players who don't have the logic to handle otherwise invalid messages.

**Reply by xesdoog** [reply]
Thank you

Yes this totally makes sense. I'm going to do some more testing. My knowledge of game hacking is close to none so I can't look look into the game itself but what I'm going to do instead is simply ask a friend to use an option from one of my Lua scripts once and then watch me use the same option and see if the particle effect becomes visible to them.

**Reply by FIying-Scotsman** [reply]
You could also look for particle effects in the scripts and note down the dict they use. That'll allow you to log the dictionary and the script to see if there's any commonalities (could be they're using only a small handful of dicts online for example).

A simple python or Lua script could search all the decompiled scripts for the PTFX native and writes it to a CSV for example.

**Reply by xesdoog** [reply]
> A simple python or Lua script could search all the decompiled scripts for the PTFX native and writes it to a CSV for example.

I could do that in Python.
I've tried the friend testing solution and it didn't work. The sound works though and even triggers a sound spam protection (it's not toxic it's just exhaust pops that sound like gunshots)
I have the 1.69 decompiled scripts from root-cause, I'll try dumping all PTFX native calls from them and see if I can figure out something.

Thank you again, I appreciate you taking the time to look into it.

**Comment by maybegreat48** [comment]
`PLAY_SOUND_FROM_ENTITY` is hardcoded not to work on player peds if that's what you're looking for. Set P1 to true when calling `REQUEST_MISSION_AUDIO_BANK` and set isNetwork for `PLAY_SOUND_FROM_COORD`, then your audio will be networked

**Reply by xesdoog** [reply]
Hello MBG, thanks for replying

The audio works and is synced Online but for players using menus this triggers a sound spam protection which blocks the sound. This makes me question if it's safe to use online or not. I'm calling `PLAY_SOUND_FROM_ENTITY()` and using the vehicle as the sound source and as you can see in this example, it gets a little bit too much if you spam it but it's not meant to be toxic towards other players


https://github.com/user-attachments/assets/dccd9964-a228-459d-ac18-b2c569af22c9
